[FSDP] Move tensor sharding logic to `FlatParamHandle`

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #81243 [FSDP] Move `_post_backward_called` to `_init_param_attributes`
* #80185 [FSDP] Clean up `_lazy_init()`
* #80087 [Easy][FSDP] Add `zero_grad()` to unit test train loop
* #80002 [FSDP] Remove `self.numel_padded_per_param` (unused)
* **#80000 [FSDP] Move tensor sharding logic to `FlatParamHandle`**
* #79984 [FSDP] Deduplicate `_orig_size` and `_unsharded_size`
* #79652 [FSDP] Introduce `FlatParamHandle`

This moves the tensor sharding logic from `FullyShardedDataParallel` to `FlatParamHandle`. In particular, `_get_shard()` and its related subroutines are moved to `FlatParamHandle` as static methods.

The motivation is to start refactoring to move the broader FSDP sharding logic in `_shard_parameters()` to `FlatParamHandle` (as a part of the multiple parameter group and possibly future pluggable sharding efforts). In other words, in follow-ups, I hope to move
https://github.com/pytorch/pytorch/blob/cd089544631d3bcad3b620865543a649e2535cbf/torch/distributed/fsdp/fully_sharded_data_parallel.py#L1444-L1447
to be part of `FlatParamHandle`.

Differential Revision: [D37726060](https://our.internmc.facebook.com/intern/diff/D37726060)

Commit:
[FSDP] Move tensor sharding logic to `FlatParamHandle`

[ghstack-poisoned]

Commit:
Update on "[FSDP] Move tensor sharding logic to `FlatParamHandle`"




[ghstack-poisoned]

rohan-varma commented:
cc: @fegin

fegin commented on torch/distributed/fsdp/flat_param.py:
```diff
@@ -432,6 +433,65 @@ def _get_shard_metadata(
         return tuple(shard_param_offsets), shard_param_indices
 
+    @staticmethod
+    def _get_shard_unpadded(
```
nit: _get_unpadded_shard seems to be a more nature naming?

fegin commented on torch/distributed/fsdp/flat_param.py:
```diff
@@ -432,6 +433,65 @@ def _get_shard_metadata(
+        tensor: Tensor,
+        rank: int,
+        world_size: int,
+    ) -> Tuple[Tensor, int]:
```
Should we consider to avoid using `Tuple` in the future but use `NamedTuple` for better readability during this code refactor?

awgu replied:
Makes sense. I will do this in follow-ups.

fegin reviewed (approved):
LGTM, just some minor nits.

Commit:
Update on "[FSDP] Move tensor sharding logic to `FlatParamHandle`"


This moves the tensor sharding logic from `FullyShardedDataParallel` to `FlatParamHandle`. In particular, `_get_shard()` and its related subroutines are moved to `FlatParamHandle` as static methods.

The motivation is to start refactoring to move the broader FSDP sharding logic in `_shard_parameters()` to `FlatParamHandle` (as a part of the multiple parameter group and possibly future pluggable sharding efforts). In other words, in follow-ups, I hope to move
https://github.com/pytorch/pytorch/blob/cd089544631d3bcad3b620865543a649e2535cbf/torch/distributed/fsdp/fully_sharded_data_parallel.py#L1444-L1447
to be part of `FlatParamHandle`.


[ghstack-poisoned]

Commit:
Update on "[FSDP] Move tensor sharding logic to `FlatParamHandle`"


This moves the tensor sharding logic from `FullyShardedDataParallel` to `FlatParamHandle`. In particular, `_get_shard()` and its related subroutines are moved to `FlatParamHandle` as static methods.

The motivation is to start refactoring to move the broader FSDP sharding logic in `_shard_parameters()` to `FlatParamHandle` (as a part of the multiple parameter group and possibly future pluggable sharding efforts). In other words, in follow-ups, I hope to move
https://github.com/pytorch/pytorch/blob/cd089544631d3bcad3b620865543a649e2535cbf/torch/distributed/fsdp/fully_sharded_data_parallel.py#L1444-L1447
to be part of `FlatParamHandle`.


[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37726060).

Commit:
Update on "[FSDP] Move tensor sharding logic to `FlatParamHandle`"


This moves the tensor sharding logic from `FullyShardedDataParallel` to `FlatParamHandle`. In particular, `_get_shard()` and its related subroutines are moved to `FlatParamHandle` as static methods.

The motivation is to start refactoring to move the broader FSDP sharding logic in `_shard_parameters()` to `FlatParamHandle` (as a part of the multiple parameter group and possibly future pluggable sharding efforts). In other words, in follow-ups, I hope to move
https://github.com/pytorch/pytorch/blob/cd089544631d3bcad3b620865543a649e2535cbf/torch/distributed/fsdp/fully_sharded_data_parallel.py#L1444-L1447
to be part of `FlatParamHandle`.

Differential Revision: [D37726060](https://our.internmc.facebook.com/intern/diff/D37726060)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37726060).

Commit:
Update on "[FSDP] Move tensor sharding logic to `FlatParamHandle`"


This moves the tensor sharding logic from `FullyShardedDataParallel` to `FlatParamHandle`. In particular, `_get_shard()` and its related subroutines are moved to `FlatParamHandle` as static methods.

The motivation is to start refactoring to move the broader FSDP sharding logic in `_shard_parameters()` to `FlatParamHandle` (as a part of the multiple parameter group and possibly future pluggable sharding efforts). In other words, in follow-ups, I hope to move
https://github.com/pytorch/pytorch/blob/cd089544631d3bcad3b620865543a649e2535cbf/torch/distributed/fsdp/fully_sharded_data_parallel.py#L1444-L1447
to be part of `FlatParamHandle`.

Differential Revision: [D37726060](https://our.internmc.facebook.com/intern/diff/D37726060)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37726060).

awgu commented:
@pytorchbot merge
